torch/csrc/autograd/python_function.h

```diff
@@ -89,9 +97,11 @@ struct THPFunction {
     std::vector<bool> is_variable_input;
     char has_freed_buffers;
 
-    // The C++ wrapper for this Python function.
-    // See a comment in THPFunction_asFunction for details about this field.
-    torch::autograd::PyFunction cdata;
+    // The actual PyFunction (in the autograd graph) that this data was
+    // saved for.  This field may be NULL (because a user can construct
+    // a THPFunction directly from Python), but when this field is non-NULL,
+    // it is guaranteed that cdata.lock()->obj == this
```

Comment:
Why can't we enforce that Python `Function` objects always have their corresponding `PyFunction`s by allocating one whenever you instantiate an object? That way we could avoid having a weak pointer, and could hold `PyFunction*` will the invariant that it would always be non-null.

Reply:
You also seem to be using the assumption that `.lock()` succeeds quite liberally in this patch, so this comment is highly confusing for me.

Reply:
> Why can't we enforce that Python Function objects always have their corresponding PyFunctions by allocating one whenever you instantiate an object?

There are a few layers to this question, so I want to address them separately.

**In the non-legacy case, why can't we allocate a PyFunction in `THPFunction_new` so that `THPFunction` is always allocated with an associated `PyFunction`?** The proximal cause is dumb: it's because `THPFunction` only holds a weak pointer to `PyFunction`, so if you allocate a `PyFunction` inside `THPFunction_new`, the only pointer that would be keeping it live is a weak pointer, and it would immediately go dead once `THPFunction_new` returns. 

"But wait, Edward!" you might say, "Doesn't `THPFunction_apply` have this problem too?" No! We keep an owning reference to `PyFunction` for the entire body of this function, and eventually we will store an OWNING reference to `PyFunction` in the `grad_fn` of the Variables we return (in `process_outputs`); that owning reference will keep PyFunction live when we return. But there's no way to setup this owning reference in the constructor `THPFunction_new`, so we have to set it up somewhere else. It's a bit finely balanced, but in the non-legacy case, as long as you don't go constructing copies of `THPFunction` outside of `apply` (it is possible to do this, but not in the public API), we maintain the invariant that every THPFunction has a corresponding PyFunction.

**In the legacy case, why can't we guarantee that there is an associated PyFunction?** The first answer is, the legacy case is dumb and we shouldn't support it. But digging deeper, I think the root problem is that `THPFunction` (the `ctx` object) is what we expose to the world as `grad_fn`, when actually we should have exposed some sort of Python binding of `PyFunction`. So there is a confusion: we treat `THPFunction` as a Python binding as `PyFunction`, but we ALSO treat it as the `ctx` object we pass in when doing autograd. These two use cases are incompatible, but right now they're smooshed together in a single object.

The suggestion you've made here is treating `THPFunction` as just a Python binding for `PyFunction`. Which is reasonable, but a much more involved change.